File: concepts/data structures/implementation/tree/bt_traversal/depth_first/typescript/bt_in_order.ts
import { BinaryNode } from "../../../bt_base/typescript/binary_tree";

function walk(curr: BinaryNode<number> | null, path: number[]): number[] { 
  if (!curr) return path;

  // recursion
  // pre
  // recurse
  walk(curr.left, path);
  path.push(curr.value);
  walk(curr.right, path);

  // post
  return path;
}

function in_order_search(head: BinaryNode<number>): number[] {
  return walk(head, []);
}

function testInOrderSearch(): void {
  const tree = new BinaryNode(
    1,
    new BinaryNode(2, new BinaryNode(4), new BinaryNode(5)),
    new BinaryNode(3),
  );
  const result = in_order_search(tree);

  console.log(`In-order: ${result.join(", ")} | Expected: 4, 2, 5, 1, 3`);
}

testInOrderSearch();

File: concepts/data structures/implementation/tree/bt_traversal/depth_first/python/bt_in_order.py
from bt_base.python.binary_tree import BinaryNode


def walk(curr, path):
    if curr is None:
        return path

    # Recursion
    # Pre
    # Recurse
    walk(curr.left, path)
    path.append(curr.value)
    walk(curr.right, path)

    # Post
    return path


def in_order_search(head):
    return walk(head, [])


def test_in_order_search():
    tree = BinaryNode(
        1,
        BinaryNode(2, BinaryNode(4), BinaryNode(5)),
        BinaryNode(3),
    )
    result = in_order_search(tree)

    print(f"In-order: {result} | Expected: [4, 2, 5, 1, 3]")


test_in_order_search()
